Baker - Enhancements - User controll over Baker #2592

bob0bob opened the pull request:
I did some searches to see if active work is being done on the Baker and I've suggest these features about 9 months ago as what I thought was a good enhancement.

If there is a procedure on new things and how to go about it. Please let me know. I didn't notice one but I manage there is one and hope this doesn't make people mad. Just trying to make the baker a nicer part of the colony.

This is an enhancement for the Baker. I've added a second gui that the user can select which recipes the baker will make. As you can see from the screen shots it shows everything recipe the baker can make. Click on the items you wish them to make. So you can control what they produce. If you only want cookies, just have the cookies the only item clicked. The baker will only make cookies. I've added Pumpkin Pie since it is a vanilla minecraft item.

So there are 4 recipes for the baker now.

If your town gets over loaded with food, you can also uncheck everything and the baker goes IDLE until there is an item that the baker is allowed to bake. You don't have to reassign them to another job anymore if you don't wish too.

Also, in this version, you have the ability to have one baker make cakes, and another make bread. Each building saves what is made in that building so you can have different buildings making different things.

There is no default recipe the baker will request if items are missing like before. They will find the first recipe that is allowed and request those items and until they receive those items, they will wait for the request to be fulfilled.

If you have more than one item checked for allowed recipes, the baker will just go down the list. The baker doesn't care if recipe 1 is fulfill-able and recipe 2 is not. If recipe 2 is the next item to be produce it will request those items and wait for it.

Unless, your building is higher than level 1. So the baker can make more than one item at a time. Say it was level 3. Which means they can be making 3 things at a time.

I've submitted all the code changes. I hope that you guys will accept the new changes to the baker. I think it still follows the scheme of MineColonies.

bob0bob committed:
Baker - Enhancements - User controll over Baker #2592

I did some searches to see if active work is being done on the Baker and I've suggest these features about 9 months ago as what I thought was a good enhancement.

If there is a procedure on new things and how to go about it. Please let me know. I didn't notice one but I manage there is one and hope this doesn't make people mad. Just trying to make the baker a nicer part of the colony.

This is an enhancement for the Baker. I've added a second gui that the user can select which recipes the baker will make. As you can see from the screen shots it shows everything recipe the baker can make. Click on the items you wish them to make. So you can control what they produce. If you only want cookies, just have the cookies the only item clicked. The baker will only make cookies. I've added Pumpkin Pie since it is a vanilla minecraft item.

So there are 4 recipes for the baker now.

If your town gets over loaded with food, you can also uncheck everything and the baker goes IDLE until there is an item that the baker is allowed to bake. You don't have to reassign them to another job anymore if you don't wish too.

Also, in this version, you have the ability to have one baker make cakes, and another make bread. Each building saves what is made in that building so you can have different buildings making different things.

There is no default recipe the baker will request if items are missing like before. They will find the first recipe that is allowed and request those items and until they receive those items, they will wait for the request to be fulfilled.

If you have more than one item checked for allowed recipes, the baker will just go down the list. The baker doesn't care if recipe 1 is fulfill-able and recipe 2 is not. If recipe 2 is the next item to be produce it will request those items and wait for it.

Unless, your building is higher than level 1. So the baker can make more than one item at a time. Say it was level 3. Which means they can be making 3 things at a time.

I've submitted all the code changes. I hope that you guys will accept the new changes to the baker. I think it still follows the scheme of MineColonies.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -90,6 +111,10 @@
     private int ticksPassed = 0;
 
+    
+    
```
javadoc please, what is this array. Describe it

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -418,11 +469,69 @@ public void putInFurnace(final BlockPos currentFurnace, final BakingProduct baki
+    {
+    	if (pos < recipesAllowed.length)
+    	{
+    		recipesAllowed[pos] = value;
```
formatting to the left

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -418,11 +469,69 @@ public void putInFurnace(final BlockPos currentFurnace, final BakingProduct baki
+    	 * list of recipes for the baker, 
+    	 * true - indicates baker is able to make recipe
+    	 */
+    	private boolean[] recipesAllowed;
```
you can copy this javadoc to the above array.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -434,6 +543,40 @@ public View(final ColonyView c, final BlockPos l)
+        	if (pos < recipesAllowed.length)
+			{
+        		recipesAllowed[pos] = value;
+			}
```
formatting

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:
```diff
@@ -278,35 +293,57 @@ private AIState kneadTheDough()
+        else
+        {
+        	setDelay(UNABLE_TO_CRAFT_DELAY);
+        	return IDLE;
```
formatting

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New  | Old  | Diff  |
| ------ |:-----:|:----:|:----:|:-----:|
| Total  | 34102 | 3740 | 2722 | +1018 |
| Errors |  593  |  26  |  17  |  +9   |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=11728&tab=Inspection&buildTypeId=bt88)

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
+	        if (copy != null)
+	        {
+	            //Wheat will be reduced by chance only (Between 3 and 6- getBuildingLevel, meaning 3-5, 3-4, 3-3, 3-2, 3-1)
+	            final int form = (getOwnBuilding().getMaxBuildingLevel() + 1) - (getOwnBuilding().getBuildingLevel() + ItemStackUtils.getSize(copy));
```
formatting

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
+	        for (final ItemStack stack : storage.getInput())
+	        {
+                list.add(stack);
+	        }
```
formatting

winex replied:
yeah, every editor should SHOW tabs, `git diff` does. not to mention that those editors should not allow mixing tabs with spaces (i use tabs ONLY for indentation in my own projects).
i personally turned off using tabs (in `mcedit` from `mc`) for working on minecolonies code

bob0bob committed:
formatting changes.

bob0bob committed:
formatting

I'm seeing what is going on.  Some code people are using "TABS" and some are using "SPACES".  So alignment is getting off because of this.

Raycoms commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:
```diff
@@ -317,42 +354,58 @@ private AIState createNewProduct()
-            {
-                copy = stack.copy();
-            }
+        	final List<ItemStack> list = new ArrayList<>();
```
the formatting here seems off

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  1  |  -1  |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  1  |  -1  |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

bob0bob committed:
formatting issues.

bob0bob committed:
more formatting issues.

bob0bob committed:
formatting changes.

LDTTeam-Buildsystem reviewed (dismissed):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34101 |  0  |  0  |  0   |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

bob0bob committed:
Merge branch 'version/1.12' into baker---pr2592

LDTTeam-Buildsystem reviewed (approved):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34103 |  2  |  0  |  +2  |
| Errors |  593  |  0  |  0  |  +0  |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=12663&tab=Inspection&buildTypeId=bt88)

LDTTeam-Buildsystem reviewed (approved):
#### Analysis Complete
##### Statistics:
|        | Count | New | Old | Diff |
| ------ |:-----:|:---:|:---:|:----:|
| Total  | 34103 |  0  |  0  |  0   |
| Errors |  593  |  0  |  0  |  0   |
##### More Information:

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -14,6 +15,9 @@
+
+import io.netty.buffer.ByteBuf;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockFurnace;
```
### Errors:
- Unused import import net.minecraft.block.BlockFurnace;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -4,6 +4,7 @@
 import com.minecolonies.api.util.BlockPosUtil;
```
### Errors:
- Unused import import com.minecolonies.api.util.BlockPosUtil;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -14,6 +15,9 @@
+import io.netty.buffer.ByteBuf;
 import net.minecraft.block.Block;
 import net.minecraft.block.BlockFurnace;
 import net.minecraft.block.state.IBlockState;
```
### Errors:
- Unused import import net.minecraft.block.state.IBlockState;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -26,9 +30,16 @@
 import org.jetbrains.annotations.NotNull;
 
-import java.util.*;
 
```
### Errors:
- Unused import import static com.minecolonies.api.util.constant.Constants.TICKS_SECOND;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -4,6 +4,7 @@
 import com.minecolonies.api.util.BlockPosUtil;
 import com.minecolonies.api.util.Log;
```
### Errors:
- Unused import import com.minecolonies.api.util.Log;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -26,9 +30,16 @@
```
### Errors:
- Unused import import net.minecraft.init.Blocks;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -26,9 +30,16 @@
 import net.minecraft.world.World;
```
### Errors:
- Unused import import net.minecraft.world.World;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -26,9 +30,16 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.Constants;
```
### Errors:
- Unused import import net.minecraftforge.common.util.Constants;

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -26,9 +30,16 @@
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.Constants;
 import net.minecraftforge.fml.common.gameevent.TickEvent;
```
### Errors:
- Unused import import net.minecraftforge.fml.common.gameevent.TickEvent;

bob0bob replied:
neede import.

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -14,6 +15,9 @@
 import com.minecolonies.coremod.entity.ai.citizen.baker.ProductState;
+import com.minecolonies.coremod.network.messages.AssignBakerRecipeMessage;
+
+import io.netty.buffer.ByteBuf;
```
### Errors:
- Unused import import io.netty.buffer.ByteBuf;

bob0bob replied:
used import

LDTTeam-Buildsystem commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBaker.java:
```diff
@@ -14,6 +15,9 @@
+import com.minecolonies.coremod.network.messages.AssignBakerRecipeMessage;
+
+import io.netty.buffer.ByteBuf;
 import net.minecraft.block.Block;
```
### Errors:
- Unused import import net.minecraft.block.Block;

bob0bob replied:
These import are required and will not compile without them.

LDTTeam-Buildsystem reviewed (approved):
#### Analysis Complete
##### Statistics:
|        | Count | New  | Old  | Diff |
| ------ |:-----:|:----:|:----:|:----:|
| Total  | 34171 | 1763 | 1695 | +68  |
| Errors |  592  |  0   |  1   | +-1  |
##### More Information:
- [Analysis - Full](https://teamcity.minecolonies.com/viewLog.html?buildId=12689&tab=Inspection&buildTypeId=bt88)

bob0bob commented:
Not sure what is going on with last check.  It says so many import not needed when they are.

winex commented:
@LDTTeam-Buildsystem, another check please, you've gone crazy! :P

winex commented:
@bob0bob btw, can you check please if your new code fixes Baker not requesting **up to 3x** Wheat if he already has only **1x** Wheat in inventory?
i think there is something wrong here: https://github.com/ldtteam/minecolonies/blob/083d7de41a7a83feaa4cd80f675b18d2ba8f1c79/src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java#L323-L346
and in your code it starts at line :360
basically, he should request **at least 2x** Wheat in this case, but he doesn't!

one can confirm this by:
- clear Warehouse, Dman and Baker from Wheat
- give Baker only 4x Wheat - he'll bake 1 bread (or only 1x Wheat)
- there will be no requests... :(

winex commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/EntityAIWorkBaker.java:
```diff
@@ -278,35 +293,57 @@ private AIState kneadTheDough()
     {
-        progress = 0;
-        final List<IItemHandler> handlers = getOwnBuilding().getHandlers();
+    	progress = 0;
```
formatting, think there's a TAB right before variable :(

winex commented on src/main/java/com/minecolonies/coremod/entity/ai/citizen/baker/BakerRecipes.java:
```diff
@@ -56,7 +59,15 @@
+
+    	
+    	
+    	final List<ItemStack> inputCake = new ArrayList<>();
```
formatting, same as below

winex commented:
yeah, that's nice guys. but what about the 1xWheat bug? sorry for not opening new issue - i just thought discussing with latest contributor on this part is better as this code weren't merged at that time

Raycoms commented:
It got fixed by this.

2018-07-19 14:52 GMT-03:00 Roman Vasiyarov <notifications@github.com>:

> yeah, that's nice guys. but what about the 1xWheat bug? sorry for not
> opening new issue - i just thought discussing with latest contributor on
> this part is better as this code weren't merged at that time
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2625#issuecomment-406361253>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y2I-5WyFbwyW7kx-tFGfsA9Crahoks5uIMdQgaJpZM4VPu5P>
> .
>

winex commented:
@Raycoms ok, i'll test it then, thanks...

Raycoms commented:
It still requests only 3 at max, but the dman brings 1 stack so he will
bring as many as he finds min 3, max 64

2018-07-19 14:55 GMT-03:00 Roman Vasiyarov <notifications@github.com>:

> @Raycoms <https://github.com/Raycoms> ok, i'll test it then, thanks...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2625#issuecomment-406362321>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y6ICIc9JF5wLoR1m9VD_kPJyL7Mgks5uIMgsgaJpZM4VPu5P>
> .
>

bob0bob commented:
I know there is changes related to requesting because the baker would not keep requesting it based on status.  I did put the call check if item is request and if not request them based on the status.  So  in the past they might request it but if it gets wiped out then it would never ask again.  In the new code it calls that routine everytime when at a certain status.

So it should be working, but yeah.  The Dman will deliver a whole stack and not just what is requested.

winex commented:
i see, and i'll test that to be sure that Baker doesn't get stuck as before.
about Dman - i hope if there is only 2x Wheat in the Warehouse (and he delivers it) - Baker would still be requesting at least +1x Wheat. that's how it **must** work!
+i hope that Baker wouldn't lag the server being in that _certain status_.

p.s. and, of course, thanks for the work - it's really cool! :)

Raycoms commented:
Yeah, he will request until he has 3

2018-07-19 15:16 GMT-03:00 Roman Vasiyarov <notifications@github.com>:

> i see, and i'll test that to be sure that Baker doesn't get stuck as
> before.
> about Dman - i hope if there is only 2x Wheat in the Warehouse (and he
> delivers it) - Baker would still be requesting at least +1x Wheat. that's
> how it *must* work!
> +i hope that Baker wouldn't lag the server being in that *certain status*.
>
> p.s. and, of course, thanks for the work - it's really cool! :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ldtteam/minecolonies/pull/2625#issuecomment-406368412>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGI9y6yp2y_Y2urr8qTUzol_3vSsZd6Iks5uIM0SgaJpZM4VPu5P>
> .
>
